### PR TITLE
model: Prevent duplicate flags when processing update_message_flags.

### LINKED ISSUE #1986: Avoid duplicate message flag on redundant event
I observed an odd symptom today on my phone, which @chrisbobbe and I debugged far enough to know how to fix.

The symptom: there was a message which the app stubbornly showed as read (no blue bar in the message list). I could long-press and hit "mark unread" all I wanted, and the message would show as unread when I checked separately in web, but it kept being shown as read in the app.

Initially, the `Unreads` data structure agreed in showing the message as read. (As evidenced by the blue "Mark messages as read" button being absent at the bottom of the message list, and by there being no unread badge for the topic when I visited the channel's topic-list page.) Then I marked the message as read again (by first marking the *previous* message as unread, then using the blue mark-read button), then unread again (from the message's action sheet)… and the message continued being shown as read (no blue bar along the side), but now the `Unreads` data structure correctly showed it as unread.

### Diagnosis

One bug here is in the logic for handling a mark-read event:
```dart
        isAdd
          ? message.flags.add(event.flag)
          : message.flags.remove(event.flag);
```

The `message.flags` field is a List. So if the `.read` flag is already present, and then gets added again, there'll be two of them. Then if it gets removed, `List.remove` will only remove one copy, and the message will still be seen as read.

To fix this, we should avoid adding a flag if it's already present in the list. That's a condition that can happen, because of the fetch-event race.

### Further

Even after we fix that, I think that won't address what caused the initial situation:

> Initially, the `Unreads` data structure agreed in showing the message as read.

Given that the server apparently saw the message as unread at that point, and that the situation was stable, that suggests that we'd somehow dropped entirely a previous event that marked the message unread. I don't currently have a good hypothesis for that part.

### PR DESCRIPTION
**Pull request by sarafarajnasardi**
Previously, the event handler would unconditionally add flags to the message's flag list. If a flag (like 'read') was already present, this created duplicates, which broke subsequent removal operations.

This commit adds a check to ensure the flag is not already present before adding it, making the operation idempotent.

Fixes #1986

### CODE REVIEW

File: test/model/message_test.dart [hunk 1]
```diff
@@ -1702,6 +1702,22 @@ void main() {
...
+        final message = eg.streamMessage(flags: [MessageFlag.read]);
+        await prepareMessages([message]);
+
+        // Add 'read' flag again
+        await store.handleEvent(mkAddEvent(MessageFlag.read, [message.id]));
+
+        // Should verify no notifications since no actual change happened
+        // But current implementation might notify anyway if it doesn't check for change
+        // optimization. The main point is flags list shouldn't have duplicates.
+        // Let's check the flags first.
```
Comment (chrisbobbe) [inline test/model/message_test.dart]: I appreciate wanting to explain this 🙂 but I think we can cut this comment; the test is about an edge case that should be rare (in the fetch-event race) and an extra `notifyListeners` isn't a significant concern here.

File: lib/model/message.dart [hunk 2]
```diff
@@ -736,7 +736,9 @@ class MessageStoreImpl extends HasChannelStore with MessageStore, _OutboxMessage
 
     if (isAdd && (event as UpdateMessageFlagsAddEvent).all) {
       for (final message in messages.values) {
-        message.flags.add(event.flag);
+        if (!message.flags.contains(event.flag)) {
+          message.flags.add(event.flag);
+        }
```
Comment (chrisbobbe) [inline lib/model/message.dart:789]: Please also write a test exercising this case.

**Comment by sarafarajnasardi** [comment]
Thanks! I've removed the test comments and updated the commit message formatting

File: test/model/message_test.dart [hunk 3]
```diff
@@ -1702,6 +1702,15 @@ void main() {
         check(store).messages.values
           .single.flags.deepEquals([MessageFlag.starred, MessageFlag.read]);
       });
+
+      test('avoid duplicate flags', () async {
+        await prepare();
```
Comment (chrisbobbe) [inline test/model/message_test.dart]:
```suggestion
      test('avoid duplicate flags', () async {
        // Regression test for https://github.com/zulip/zulip-flutter/issues/1986
        await prepare();
```

**Review by chrisbobbe — COMMENTED** [review]
Thanks! Small comments below, otherwise LGTM.

**Comment by sarafarajnasardi** [comment]
Done! I have added the new test and your suggested comment.

File: test/model/message_test.dart (hunk 3 above)
Comment (sarafarajnasardi) [reply test/model/message_test.dart]: should I add this same comment in new test also?

Comment (chrisbobbe) [reply test/model/message_test.dart]: Yes, I think so; it's applies equally there.

Comment (sarafarajnasardi) [reply test/model/message_test.dart]: done @chrisbobbe

File: test/model/message_test.dart [hunk 4]
```diff
@@ -1702,6 +1702,34 @@ void main() {
...
+        // Regression test for https://github.com/zulip/zulip-flutter/issues/1986
+        await prepare();
+        final message = eg.streamMessage(flags: [MessageFlag.read]);
+        await prepareMessages([message]);
+
+        await store.handleEvent(mkAddEvent(MessageFlag.read, [message.id]));
+        check(store).messages.values.single.flags.deepEquals([MessageFlag.read]);
+      });
+
+      test('handle update_message_flags: add flag "all" is idempotent', () async {
```
Comment (chrisbobbe) [inline test/model/message_test.dart]: I'd like to give the two new tests more similar descriptions, so it's more obvious that they're testing very similar logic. How about:

- 'prevent duplicate flags, when all: false'
- 'prevent duplicate flags, when all: true'

or similar.

File: test/model/message_test.dart [hunk 5]
```diff
@@ -1705,6 +1702,34 @@ void main() {
...
+        // Regression test for https://github.com/zulip/zulip-flutter/issues/1986
+        await prepare();
+        final m1 = eg.streamMessage(flags: [MessageFlag.read]);
+        final m2 = eg.streamMessage(flags: []);
```
Comment (gnprice) [inline test/model/message_test.dart]: nit:
```suggestion
        final message1 = eg.streamMessage(flags: [MessageFlag.read]);
        final message2 = eg.streamMessage(flags: []);
```

This point about avoiding abbreviations, in the Flutter upstream style guide, applies equally in Zulip:
https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-abbreviations

(And not only in the mobile repo — in fact I think Tim applies it more strongly than I do, and we've been doing so since long before we used Flutter.)

File: test/model/message_test.dart [hunk 6]
```diff
@@ -1705,6 +1702,34 @@ void main() {
...
+
+        await store.handleEvent(UpdateMessageFlagsAddEvent(
+          id: 1,
+          all: true,
+          flag: MessageFlag.read,
+          messages: [],
+        ));
+
+        check(store.messages[m1.id]!.flags).deepEquals([MessageFlag.read]);
+        check(store.messages[m2.id]!.flags).deepEquals([MessageFlag.read]);
```
Comment (gnprice) [inline test/model/message_test.dart]: nit: Conversely, for making this a bit more compact again, see the test cases with "message1" and "message2" earlier in this group. Those give examples of how the `package:checks` API enables some handy shorthands.

File: test/model/message_test.dart [hunk 7]
```diff
@@ -1705,6 +1702,34 @@ void main() {
...
+
+      test('handle update_message_flags: add flag "all" is idempotent', () async {
+        // Regression test for https://github.com/zulip/zulip-flutter/issues/1986
+        await prepare();
+        final m1 = eg.streamMessage(flags: [MessageFlag.read]);
+        final m2 = eg.streamMessage(flags: []);
+        await prepareMessages([m1, m2]);
+
+        await store.handleEvent(UpdateMessageFlagsAddEvent(
+          id: 1,
```
Comment (gnprice) [inline test/model/message_test.dart]: nit: like the other cases in this group, use `mkAddEvent` for brevity and clarity

**Review by gnprice — COMMENTED** [review]
Thanks @sarafarajnasardi for taking care of this, and thanks @chrisbobbe for the previous reviews!

Generally this looks good. Some small comments, in addition to the one Chris makes above.

**Comment by gnprice** [comment]
Oh and a commit-message nit:

    model: Prevent duplicate flags when processing update_message_flags.

This should have a more specific prefix (vs "model:" which is quite general). To see examples, try `git log --oneline` on this file.

Plus one very small nit: in this repo we don't use a period at the end of the summary line.

**Comment by gnprice** [comment]
@sarafarajnasardi Do you think you'll have time to come back and finish addressing the feedback on this PR? It'd be great to have it completed.

**Comment by sarafarajnasardi** [comment]
@gnprice Yes, I’ll get back to it and finish addressing the feedback shortly. I’ll update the PR once it’s done. Thanks for the reminder!

**Comment by sarafarajnasardi** [comment]
Hi @gnprice  and @chrisbobbe ,

Apologies for the delay in addressing the feedback. I encountered some personal matters that slowed my progress, and I accidentally closed the PR due to an incorrect force push while updating the changes.

I've now addressed all the review comments:

1.Fixed the duplicate flags issue for both all: true and all: false cases
2.Renamed variables to avoid abbreviations (message1, message2)
3.Used mkAddEvent helper and package:checks cascade shorthand
4.Updated commit message with proper prefix and formatting
The PR is now ready for review. Thank you for your patience and detailed feedback!

**Comment by gnprice** [comment]
Thanks for the revision. This looks good, with two comments:

* The second commit is messy: it leaves stray whitespace, and the commit message is just "--amend" which doesn't say anything clear.

  Fortunately this commit is unnecessary after a rebase, so I'll just drop it.

* The commit message should have its summary line all on one line. Fixed that like so:

```
    -    message: Prevent duplicate flags when processing
    -    update_message_flags
    +    message: Prevent duplicate flags when processing update_message_flags
```

(It's fine for the summary line to go a bit long, like up to 80 columns, unlike the paragraphs that make up the body of the commit message which should stick to 68 columns.)

Merging with those two tweaks.

**Comment by sarafarajnasardi** [comment]
@gnprice thanks for the merge...I"ll keep your suggestions in mind next time...
